cmake/FindHAL.cmake

```diff
@@ -1,251 +1,84 @@
-    adc comp cordic crc crs dac dlyb dma dma2d exti fmac fmc gpio i2c icache lpgpio lptim lpuart opamp pka pwr rcc 
-    rng rtc sdmmc spi tim ucpd usart usb utils
-)
+foreach(COMP ${HAL_REQUESTED_FAMILIES})
+    string(TOLOWER ${COMP} COMP_L)
+    string(TOUPPER ${COMP} COMP_U)
+    string(REGEX MATCH "^STM32([FGHLMUW]P?[0-9BL])([0-9A-Z][0-9M][A-Z][0-9A-Z])?_?(M0PLUS|M4|M7)?.*$" COMP_U ${COMP_U})
+     if(CMAKE_MATCH_1)
+         set(FAMILY ${CMAKE_MATCH_1})
+         list(APPEND HAL_FIND_COMPONENTS_FAMILIES ${COMP_U})
+     else()
+         message(FATAL_ERROR "\"${COMP}\" is not a valid FAMILY")
+     endif()
 
-set(HAL_DRIVERS_WB
-    adc comp cortex crc cryp dma exti flash gpio hsem 
-    i2c ipcc irda iwdg lcd lptim pcd pka pwr qspi rcc 
-    rng rtc sai smartcard smbus spi tim tsc uart usart wwdg
-)
-set(HAL_EX_DRIVERS_WB
-    adc crc cryp dma flash i2c pcd pwr rcc
-    rtc sai smartcard spi tim uart usart
-)
-set(HAL_LL_DRIVERS_WB
-    adc comp crc crs dma exti gpio i2c lptim lpuart pka pwr rcc 
-    rng rtc spi tim usart usb utils
-)
+    string(TOLOWER ${FAMILY} FAMILY_L)
+    find_path(HAL_${FAMILY}_PATH
+        NAMES Inc/stm32${FAMILY_L}xx_hal.h
+        PATHS "${STM32_HAL_${FAMILY}_PATH}" "${STM32_CUBE_${FAMILY}_PATH}/Drivers/STM32${FAMILY}xx_HAL_Driver"
+        NO_DEFAULT_PATH
+        )
 
-set(HAL_DRIVERS_WL
-    adc comp cortex crc cryp dac dma exti flash gpio gtzc hsem 
-    i2c i2s ipcc irda iwdg lptim pka pwr rcc 
-    rng rtc smartcard smbus spi subghz tim uart usart wwdg
-)
-set(HAL_EX_DRIVERS_WL
-    adc crc cryp dma flash i2c pwr rcc
-    rng rtc smartcard spi tim uart usart
-)
-set(HAL_LL_DRIVERS_WL
-    adc comp crc dac dma exti gpio i2c lptim lpuart pka pwr rcc 
-    rng rtc spi tim usart utils
-)
+    get_list_hal_drivers(HAL_DRIVERS_${FAMILY} ${HAL_${FAMILY}_PATH} "hal")
+    get_list_hal_drivers(HAL_LL_DRIVERS_${FAMILY} ${HAL_${FAMILY}_PATH} "ll")
+    list(APPEND HAL_DRIVERS ${HAL_DRIVERS_${FAMILY}})
+    list(APPEND HAL_LL_DRIVERS ${HAL_LL_DRIVERS_${FAMILY}})
 
-foreach(FAMILY_SUFFIX ${STM32_SUPPORTED_FAMILIES_SHORT_NAME})
-    list(APPEND HAL_DRIVERS ${HAL_DRIVERS_${FAMILY_SUFFIX}})
-    list(APPEND HAL_LL_DRIVERS ${HAL_LL_DRIVERS_${FAMILY_SUFFIX}})
+    message(TRACE "FindHAL: append COMP ${COMP} to HAL_FIND_COMPONENTS_FAMILIES")
 endforeach()
+
 list(REMOVE_DUPLICATES HAL_DRIVERS)
 list(REMOVE_DUPLICATES HAL_LL_DRIVERS)
 
-foreach(COMP ${HAL_FIND_COMPONENTS})
+#Checking the componenet requested on the find_package(HAL <params>) such as uart, ll_adc...
+foreach(COMP ${HAL_FIND_COMPONENTS_UNHANDLED})
     string(TOLOWER ${COMP} COMP_L)
     string(TOUPPER ${COMP} COMP_U)
     
-    string(REGEX MATCH "^STM32([FGHLMUW]P?[0-9BL])([0-9A-Z][0-9M][A-Z][0-9A-Z])?_?(M0PLUS|M4|M7)?.*$" COMP_U ${COMP_U})
-    if(CMAKE_MATCH_1)
-        list(APPEND HAL_FIND_COMPONENTS_FAMILIES ${COMP})
-        message(TRACE "FindHAL: append COMP ${COMP} to HAL_FIND_COMPONENTS_FAMILIES")
-        continue()
-    endif()
```

Comment:
I cant comment in PR bellow this line but please have a look at next occurences of `HAL_FIND_COMPONENTS_FAMILIES` it might explain you are breaking the multicore families tests + notice the management of empty list to manage all families. I think your work is not finished